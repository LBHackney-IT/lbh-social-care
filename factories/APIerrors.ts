import { Factory } from 'fishery';

import { ErrorAPI } from 'types';

export const errorFactory = Factory.define<ErrorAPI>(() => ({
  response: {
    data: {},
    status: 404,
    statusText: 'error',
    headers: {},
    config: {},
  },
  name: 'error',
  message: 'error',
  config: {},
  isAxiosError: false,
  toJSON: () => ({}),
}));

export const mockedAPIerror = errorFactory.build();
export const mockedAPIservererror = errorFactory.build({
  response: {
    status: 500,
  },
});
