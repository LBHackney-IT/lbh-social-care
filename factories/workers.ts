import { Factory } from 'fishery';

import { Worker } from 'types';

export const userFactory = Factory.define<Worker>(({ sequence }) => ({
  id: sequence,
  email: 'foo.bar@hackney.gov.uk',
  firstName: 'Foo',
  lastName: 'Bar',
  role: 'aliquam',
  allocationCount: 0,
  teams: [],
}));

export const mockedWorker = userFactory.build();
export const mockedWorkers = [userFactory.build()];
