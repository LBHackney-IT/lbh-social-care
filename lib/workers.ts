import axios from 'axios';

import { Worker } from 'types';

const { ENDPOINT_API, AWS_KEY } = process.env;

const headersWithKey = {
  'x-api-key': AWS_KEY,
};

export const getWorkers = async (
  params?: Record<string, unknown>
): Promise<Worker[]> => {
  const { data } = await axios.get(`${ENDPOINT_API}/workers`, {
    headers: headersWithKey,
    params,
  });
  return data;
};

export const getWorker = async (
  id: number,
  params?: Record<string, unknown>
): Promise<Worker> => {
  const { data } = await axios.get(`${ENDPOINT_API}/workers`, {
    headers: headersWithKey,
    params: { id, ...params },
  });
  return data;
};
