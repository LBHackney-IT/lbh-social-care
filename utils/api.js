import { getQueryString } from 'utils/urls';

export const getInfiniteKey = (baseUrl, key, params) => (
  pageIndex,
  previousPageData
) => {
  if (previousPageData && !previousPageData[key]) return null;
  if (pageIndex === 0)
    return `${baseUrl}${params ? `?${getQueryString(params)}` : ''}`;
  return `${baseUrl}?${getQueryString({
    ...params,
    cursor: previousPageData.nextCursor,
  })}`;
};
