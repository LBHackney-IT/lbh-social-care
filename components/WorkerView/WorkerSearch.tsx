import Button from 'components/Button/Button';
import { EmailInput } from 'components/Form';
import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import { useWorker } from 'utils/api/workers';
import WorkerRecap from 'components/WorkerView/WorkerRecap';
import AllocatedCases from 'components/AllocatedCases/AllocatedCases';
import Spinner from 'components/Spinner/Spinner';

interface FormValues {
  email: string;
}

const WorkerSearch = (): React.ReactElement | null => {
  const [params, setParams] = useState<FormValues>();
  const { data } = useWorker(params);
  const { register, handleSubmit, errors } = useForm();
  if (params && !data) {
    return <Spinner />;
  }
  return (
    <div>
      <form
        role="form"
        onSubmit={handleSubmit((formData: FormValues) => setParams(formData))}
      >
        <div
          className="govuk-grid-row"
          style={{
            display: 'flex',
            alignItems: 'flex-end',
            justifyContent: 'center',
          }}
        >
          <div className="govuk-grid-column-one-half">
            <EmailInput
              error={errors.email}
              name="email"
              hint="Email address must contain @hackney.gov.uk"
              label="Worker's email address:"
              labelSize="s"
              placeholder="Email"
              rules={{ required: 'Enter a valid worker email' }}
              register={register}
            />
          </div>
          <div className="govuk-grid-column-one-half">
            <Button label={'Search'} type="submit" className="govuk-button" />
          </div>
        </div>
      </form>
      {data?.[0] && (
        <>
          <WorkerRecap {...data[0]} />{' '}
          <h2 className="gov-weight-lighter">Current Allocations</h2>
          <AllocatedCases id={data?.[0].id}></AllocatedCases>{' '}
        </>
      )}
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <div>
          <a
            href="/"
            className="govuk-link"
            onClick={() => (window.location.href = '/')}
            style={{
              lineHeight: '2.5rem',
            }}
          >
            Cancel
          </a>
        </div>
        <Button
          wideButton
          className="govuk-!-margin-left-1"
          label="Update worker's details"
          type="submit"
          onClick={() =>
            (window.location.href = '/workers/edit/?emailAddress=')
          }
        />
      </div>
    </div>
  );
};

export default WorkerSearch;
