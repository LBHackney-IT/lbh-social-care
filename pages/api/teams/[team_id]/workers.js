import * as HttpStatus from 'http-status-codes';

import { getWorkers } from 'utils/server/workers';
import { isAuthorised } from 'utils/auth';

export default async (req, res) => {
  const user = isAuthorised(req);
  if (!user) {
    return res
      .status(HttpStatus.UNAUTHORIZED)
      .json({ message: 'Auth cookie missing.' });
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getWorkers(req.query);
        res.status(HttpStatus.OK).json(data);
      } catch (error) {
        console.error('Workers get error:', error?.response?.data);
        error?.response?.status === HttpStatus.NOT_FOUND
          ? res
              .status(HttpStatus.NOT_FOUND)
              .json({ message: 'Workers Not Found' })
          : res
              .status(HttpStatus.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the Workers' });
      }
      break;

    default:
      res
        .status(HttpStatus.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};
