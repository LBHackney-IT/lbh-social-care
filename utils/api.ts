import { getQueryString } from 'utils/urls';

interface PageData {
  nextCursor?: string;
  [key: string]: unknown;
}

export const getInfiniteKey = (
  baseUrl: string,
  key: string,
  params: Record<string, unknown>
) => (pageIndex: number, previousPageData: PageData): string | null => {
  if (previousPageData && !previousPageData[key]) return null;
  if (pageIndex === 0)
    return `${baseUrl}${params ? `?${getQueryString(params)}` : ''}`;
  return `${baseUrl}?${getQueryString({
    ...params,
    cursor: previousPageData.nextCursor,
  })}`;
};
