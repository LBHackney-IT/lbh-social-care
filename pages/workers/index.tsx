import AllocatedCases from 'components/AllocatedCases/AllocatedCases';
import Seo from 'components/Layout/Seo/Seo';
import WorkerSearch from 'components/WorkerView/WorkerSearch';
import React from 'react';
import { useRouter } from 'next/router';
// import WorkerRecap from 'components/WorkerView/WorkerRecap';

const SearchWorkerPage = (): React.ReactElement => (
  <>
    <div>
      <Seo title="Search Worker" />
      <h1 className="govuk-fieldset__legend--l gov-weight-lighter govuk-expand-title">
        {' '}
        Manage Workers{' '}
      </h1>
      <p className="govuk-body">
        Before creating or updating a worker in this system, you will need to
        check if their email address already exists in the Social Care System.
      </p>
      <p className="govuk-body">
        Workers also need to be added to an appropriate Google Group which will
        give them the correct permissions to use this system. If this has not
        already been done, contact XXXX
      </p>
    </div>
    <WorkerSearch />
  </>
);

export default SearchWorkerPage;
