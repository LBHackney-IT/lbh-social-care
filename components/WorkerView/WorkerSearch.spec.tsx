import { render } from '@testing-library/react';
import WorkerSearch from './WorkerSearch';
import * as useWorkers from 'utils/api/workers';
import { mockedWorker } from 'factories/workers';
import { mockedAPIerror } from 'factories/APIerrors';

jest.mock('components/Spinner/Spinner', () => () => 'MockedSpinner');

describe(`WorkerSearch`, () => {
  it('should display search results', async () => {
    jest.spyOn(useWorkers, 'useWorker').mockImplementation(() => ({
      data: [mockedWorker],
      revalidate: jest.fn(),
      mutate: jest.fn(),
      isValidating: false,
    }));
    const { asFragment } = render(<WorkerSearch />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should display the error', async () => {
    jest.spyOn(useWorkers, 'useWorker').mockImplementation(() => ({
      error: mockedAPIerror,
      revalidate: jest.fn(),
      mutate: jest.fn(),
      isValidating: false,
    }));
    const { asFragment } = render(<WorkerSearch />);
    expect(asFragment()).toMatchSnapshot();
  });
});
