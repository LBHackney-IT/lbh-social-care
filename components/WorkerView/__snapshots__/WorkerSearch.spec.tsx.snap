// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkerSearch should display search results 1`] = `
<DocumentFragment>
  <div>
    <form
      role="form"
    >
      <div
        class="govuk-grid-row"
        style="display: flex; align-items: flex-end; justify-content: center;"
      >
        <div
          class="govuk-grid-column-one-half"
        >
          <div
            class="lbh-form-group govuk-form-group"
          >
            <label
              class="lbh-label"
              for="email"
            >
              Worker's email address: 
            </label>
            <span
              class="lbh-hint govuk-hint "
              id="email-hint"
            >
              Email address must contain @hackney.gov.uk
            </span>
            <input
              aria-describedby="email-hint"
              class="lbh-input govuk-input"
              data-testid="email"
              id="email"
              name="email"
              placeholder="Email"
              type="email"
            />
          </div>
        </div>
        <div
          class="govuk-grid-column-one-half"
        >
          <button
            class="lbh-button govuk-button govuk-button"
            data-module="govuk-button"
            type="submit"
          >
            Search
          </button>
        </div>
      </div>
    </form>
    <div>
      <h2
        class="gov-weight-lighter"
      >
        Worker's details found in system
      </h2>
      <p
        class="govuk-body"
      >
        Name: Foo Bar
      </p>
      <p
        class="govuk-body"
      >
        Role: aliquam
      </p>
      <p
        class="govuk-body"
      >
        Teams: 
      </p>
    </div>
    <h2
      class="gov-weight-lighter"
    >
      Current Allocations
    </h2>
    <span
      class="govuk-error-message"
    >
      <span
        class="govuk-visually-hidden"
      >
        Error:
      </span>
       Oops an error occurred
    </span>
    <div
      style="display: flex; justify-content: space-between;"
    >
      <div>
        <a
          class="lbh-link"
          href="/"
        >
          Cancel
        </a>
      </div>
      <button
        class="lbh-button govuk-button govuk-!-margin-left-1"
        data-module="govuk-button"
        style="min-width: 200px;"
      >
        Update worker's details
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`WorkerSearch should display the error 1`] = `
<DocumentFragment>
  <div>
    <form
      role="form"
    >
      <div
        class="govuk-grid-row"
        style="display: flex; align-items: flex-end; justify-content: center;"
      >
        <div
          class="govuk-grid-column-one-half"
        >
          <div
            class="lbh-form-group govuk-form-group"
          >
            <label
              class="lbh-label"
              for="email"
            >
              Worker's email address: 
            </label>
            <span
              class="lbh-hint govuk-hint "
              id="email-hint"
            >
              Email address must contain @hackney.gov.uk
            </span>
            <input
              aria-describedby="email-hint"
              class="lbh-input govuk-input"
              data-testid="email"
              id="email"
              name="email"
              placeholder="Email"
              type="email"
            />
          </div>
        </div>
        <div
          class="govuk-grid-column-one-half"
        >
          <button
            class="lbh-button govuk-button govuk-button"
            data-module="govuk-button"
            type="submit"
          >
            Search
          </button>
        </div>
      </div>
    </form>
    <h3>
      Worker's email address not found
    </h3>
    <p
      class="govuk-body govuk-!-margin-top-5"
    >
      The email address is not in the system.
    </p>
    <p
      class="govuk-body govuk-!-margin-top-5"
    >
      You can search again on a different email address
    </p>
    <p
      class="govuk-body govuk-!-margin-top-5"
    >
      Please be sure the email address is correct before you use it to 
      <a
        class="govuk-link"
        href="/workers/add"
      >
        create a new worker in the system
      </a>
    </p>
  </div>
</DocumentFragment>
`;
